core/iwasm/aot/aot_runtime.c

```diff
@@ -3059,14 +3049,96 @@ aot_dump_call_stack(WASMExecEnv *exec_env)
         frame.module_offset = 0;
         frame.func_index = cur_frame->func_index;
         frame.func_offset = 0;
+        frame.func_name_wp =
+            get_func_name_from_index(module_inst, cur_frame->func_index);
 
         if (!bh_vector_append(module_inst->frames.ptr, &frame)) {
             bh_vector_destroy(module_inst->frames.ptr);
-            return;
+            return false;
         }
 
         cur_frame = cur_frame->prev_frame;
     }
+
+    return true;
+}
+
+#define PRINT_OR_DUMP()                                                \
+    do {                                                               \
+        if (print) {                                                   \
+            os_printf("%s", line_buf);                                 \
+        }                                                              \
+        else if (buf) {                                                \
+            uint32 remain_len = len - total_len;                       \
+            uint32 string_len =                                        \
+                snprintf(buf + total_len, remain_len, "%s", line_buf); \
+            if (string_len >= remain_len) {                            \
+                /* Buffer full */                                      \
+                return len;                                            \
+            }                                                          \
+            total_len += string_len;                                   \
+        }                                                              \
+        else {                                                         \
+            total_len += strlen(line_buf);                             \
+        }                                                              \
+    } while (0)
+
+uint32
+aot_dump_call_stack(WASMExecEnv *exec_env, bool print, char *buf, uint32 len)
+{
+    AOTModuleInstance *module_inst = (AOTModuleInstance *)exec_env->module_inst;
+    uint32 n = 0, total_len = 0, total_frames;
+    /* reserve 256 bytes for line buffer, any line longer than 256 bytes
+     * will be truncated */
+    char line_buf[256];
+
+    if (!module_inst->frames.ptr) {
+        return 0;
+    }
+
+    total_frames = bh_vector_size(module_inst->frames.ptr);
+    if (total_frames == 0) {
+        return 0;
+    }
+
+    snprintf(line_buf, sizeof(line_buf), "\n");
+    PRINT_OR_DUMP();
+
+    while (n < total_frames) {
+        WASMCApiFrame frame = { 0 };
+        uint32 line_length, i;
+
+        if (!bh_vector_get(module_inst->frames.ptr, n, &frame)) {
+            return 0;
```

Comment:
There are some lines printed already, is it reasonable to return 0?

Reply:
0 means error here, even if some lines has been dumped to buffer, the content is not complete, so we return 0 to indicate that the content in buf may be not valid.

Reply:
OK, got it, thanks.